src/hw_6/hw_6.ts

```diff
@@ -0,0 +1,172 @@
+enum EmployeeStatus {
+    Active,
+    Inactive,
+    OnLeave
+}
+
+interface PreHiredEmployee {
+    firstName: string;
+    lastName: string;
+    salary: number;
+    bankAccountNumber: string;
+}
+
+interface Employee extends PreHiredEmployee {
+    paymentInfo: string;
+    status: EmployeeStatus;
+    department: Department;
+}
+
+interface Department {
+    name: string;
+    domain: string;
+    employees: Employee[];
+    budget: Budget;
+
+    calculateBalance(): number;
+    addEmployee(employee: Employee): void;
+    removeEmployee(employee: Employee): void;
+    isEmployeeInDepartment(employee: Employee): boolean;
+}
+
+interface Accounting extends Department {
+    processPayroll(): void;
+    paySalary(employee: Employee | PreHiredEmployee): void;
+}
+
+interface Budget {
+    debit: number;
+    credit: number;
+}
+
+abstract class BaseDepartment implements Department {
+    public employees: Employee[] = [];
+    public budget: Budget = { debit: 0, credit: 0 };
+
+    constructor(public name: string, public domain: string) {}
+
+    abstract calculateBalance(): number;
+
+    addEmployee(employee: Employee): void {
+        this.employees.push(employee);
+    }
+
+    removeEmployee(employee: Employee): void {
+        this.employees = this.employees.filter(e => e !== employee);
+    }
+
+    isEmployeeInDepartment(employee: Employee): boolean {
+        return this.employees.includes(employee);
+    }
+}
+
+function isPreHired(employee: PreHiredEmployee | Employee): employee is PreHiredEmployee {
+    return (employee as Employee).department === undefined;
+}
+
+function isActiveEmployee(employee: Employee): boolean {
+    return employee.status === EmployeeStatus.Active;
+}
+
+function isEmployee(employee: PreHiredEmployee | Employee): employee is Employee {
+    return (employee as Employee).department !== undefined && (employee as Employee).paymentInfo !== undefined;
+}
+
+class AccountingDepartment extends BaseDepartment implements Accounting {
+    constructor() {
+        super("Accounting", "Finance");
+    }
+
+    calculateBalance(): number {
+        return this.budget.debit - this.budget.credit;
+    }
+
+    processPayroll(): void {
+        this.employees.forEach(employee => {
+            if (isActiveEmployee(employee)) {
+                this.paySalary(employee);
+            }
+        });
+        console.log("Payroll processed for active employees.");
+    }
+
+    paySalary(employee: Employee | PreHiredEmployee): void {
+        if (isEmployee(employee)) {
+            const message = isActiveEmployee(employee)
+                ? `Paying internal salary for active employee ${employee.firstName} ${employee.lastName}`
+                : `${employee.firstName} ${employee.lastName} is not eligible for payroll.`;
+            
+            console.log(message);
+            return;
+        }
+        
+        if (isPreHired(employee)) {
```

Comment:
In case you have two options - Employee and PreHiredEmployee - it is enough to perform one check and then use the "else" case. However, your implementation is acceptable as well.